File: packages/core/src/tiles.ts
/**
 * Tile type registry.
 *
 * Numeric codes are the serialization contract (typed arrays, matrix snapshots,
 * agent action payloads). Existing codes NEVER renumber — new types are always
 * appended. Phase 2 adds roads, R/C/I zoning, and power/water infrastructure.
 */
export const TILE_TYPES = {
  WATER: 0,
  SAND: 1,
  GRASS: 2,
  FOREST: 3,
  STONE: 4,
  DIRT: 5,
  // ── Phase 2: infrastructure & zoning ──
  ROAD: 6,
  RESIDENTIAL: 7,
  COMMERCIAL: 8,
  INDUSTRIAL: 9,
  POWER_PLANT: 10,
  WATER_TOWER: 11,
  // ── Phase 2.5: rail network ──
  RAIL: 12,
} as const;

export type TileType = (typeof TILE_TYPES)[keyof typeof TILE_TYPES];

export const TILE_NAMES: Record<TileType, string> = {
  [TILE_TYPES.WATER]: 'water',
  [TILE_TYPES.SAND]: 'sand',
  [TILE_TYPES.GRASS]: 'grass',
  [TILE_TYPES.FOREST]: 'forest',
  [TILE_TYPES.STONE]: 'stone',
  [TILE_TYPES.DIRT]: 'dirt',
  [TILE_TYPES.ROAD]: 'road',
  [TILE_TYPES.RESIDENTIAL]: 'residential',
  [TILE_TYPES.COMMERCIAL]: 'commercial',
  [TILE_TYPES.INDUSTRIAL]: 'industrial',
  [TILE_TYPES.POWER_PLANT]: 'power_plant',
  [TILE_TYPES.WATER_TOWER]: 'water_tower',
  [TILE_TYPES.RAIL]: 'rail',
};

export const TILE_PALETTE: Record<TileType, string> = {
  [TILE_TYPES.WATER]: '#2a5f8f',
  [TILE_TYPES.SAND]: '#d8c48a',
  [TILE_TYPES.GRASS]: '#5d9e45',
  [TILE_TYPES.FOREST]: '#3a7a36',
  [TILE_TYPES.STONE]: '#8f969d',
  [TILE_TYPES.DIRT]: '#8a6d4b',
  [TILE_TYPES.ROAD]: '#3a3f45',
  [TILE_TYPES.RESIDENTIAL]: '#d9c9a3',
  [TILE_TYPES.COMMERCIAL]: '#d9b45a',
  [TILE_TYPES.INDUSTRIAL]: '#c98a5e',
  [TILE_TYPES.POWER_PLANT]: '#a55eea',
  [TILE_TYPES.WATER_TOWER]: '#45aaf2',
  [TILE_TYPES.RAIL]: '#4a4030',
};

export const TILE_TYPE_COUNT = Object.keys(TILE_TYPES).length;

export function tileName(type: TileType): string {
  return TILE_NAMES[type] ?? `unknown(${type})`;
}

/** Inverse of TILE_NAMES: name → numeric code (for agent-readable legends). */
export function buildLegend(): Record<string, number> {
  const legend: Record<string, number> = {};
  for (const [code, name] of Object.entries(TILE_NAMES)) {
    legend[name] = Number(code);
  }
  return legend;
}

File: rust/autopolis-core/src/tiles.rs
//! Tile type registry.
//!
//! Numeric codes are the serialization contract (typed arrays, matrix
//! snapshots, agent action payloads). Existing codes NEVER renumber — new
//! types are always appended. Codes match tiles.ts exactly:
//! WATER 0 ... RAIL 12.

pub const TILE_TYPE_WATER: u8 = 0;
pub const TILE_TYPE_SAND: u8 = 1;
pub const TILE_TYPE_GRASS: u8 = 2;
pub const TILE_TYPE_FOREST: u8 = 3;
pub const TILE_TYPE_STONE: u8 = 4;
pub const TILE_TYPE_DIRT: u8 = 5;
pub const TILE_TYPE_ROAD: u8 = 6;
pub const TILE_TYPE_RESIDENTIAL: u8 = 7;
pub const TILE_TYPE_COMMERCIAL: u8 = 8;
pub const TILE_TYPE_INDUSTRIAL: u8 = 9;
pub const TILE_TYPE_POWER_PLANT: u8 = 10;
pub const TILE_TYPE_WATER_TOWER: u8 = 11;
pub const TILE_TYPE_RAIL: u8 = 12;

pub const TILE_TYPE_COUNT: usize = 13;

pub const TILE_NAMES: [&str; TILE_TYPE_COUNT] = [
    "water",
    "sand",
    "grass",
    "forest",
    "stone",
    "dirt",
    "road",
    "residential",
    "commercial",
    "industrial",
    "power_plant",
    "water_tower",
    "rail",
];

/// Color palette — unused by the simulation itself, kept for fidelity.
pub const TILE_PALETTE: [&str; TILE_TYPE_COUNT] = [
    "#2f6f9f",
    "#d8c48a",
    "#6fae4f",
    "#3f7d3a",
    "#9aa1a8",
    "#8a6d4b",
    "#4a5058",
    "#6fa8dc",
    "#e6c15c",
    "#c98a5e",
    "#a55eea",
    "#45aaf2",
    "#6b5b3e",
];

pub fn tile_name(type_code: u8) -> String {
    if (type_code as usize) < TILE_TYPE_COUNT {
        TILE_NAMES[type_code as usize].to_string()
    } else {
        format!("unknown({type_code})")
    }
}
